metatensor-core/src/labels.rs

```diff
@@ -96,14 +96,21 @@ impl LabelValue {
     }
 }
 
-type DefaultHasher = std::hash::BuildHasherDefault<ahash::AHasher>;
+// Labels uses `AHash` instead of the default hasher in std since `AHash` is
+// much faster and we don't need the cryptographic strength hash from std.
+type AHashHasher = std::hash::BuildHasherDefault<ahash::AHasher>;
+
+// Use a small vec to store Labels entries in the `positions`. This helps by
+// removing heap allocations in the most common case (fewer than 8 dimensions in
+// the Labels), while still allowing the Labels to contains many dimensions.
+type LabelsEntry = SmallVec<[LabelValue; 8]>;
```

Comment:
Let me understand - does this mean the minimum size of a labelsentry  is 8 ints? Does this have memory implications?

Reply:
Yes, this will make the LabelsEntry at least 33 bytes (8 ints + 1 discriminant to know if the data is on stack or the heap). Keep in mind that LabeslEntry is only used inside the `positions` HashMap, and not the `values` array.

In the worst case (Labels with 1 dimension), this means we'll use `size * 4` (in the values) + `size * (33 + 2)` (in the `positions` HashMap, 33 bytes per labels entry and 2 bytes for the position index) bytes of storage.

So for 100000 entries, this will take 3.7 MiB of RAM, which feels fine to me. When running on GPU, only the `values` are duplicated in GPU memory, so that would be an extra 390kiB of VRAM.

Finally, I'm looking into creating the HashMap only when required (i.e. when the user calls a function that needs it), which would further reduce the memory footprint.

Reply:
Understood. Maybe one could make the size a #DEFINE so that one can set it, but seems overkill.